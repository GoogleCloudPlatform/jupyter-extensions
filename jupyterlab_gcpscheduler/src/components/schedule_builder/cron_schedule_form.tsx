/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  css,
  FieldError,
  LearnMoreLink,
  TextInput,
} from 'gcp_jupyterlab_shared';
import * as React from 'react';

import { OnScheduleChange } from './schedule_builder';

const SCHEDULE_LINK =
  'https://cloud.google.com/scheduler/docs/configuring/cron-job-schedules';

interface SubFormProps {
  schedule: string;
  onScheduleChange: OnScheduleChange;
}

export function CronScheduleBuilder(props: SubFormProps) {
  return (
    <div>
      <TextInput
        label="Schedule"
        name="schedule"
        value={props.schedule}
        hasError={!props.schedule}
        onChange={e => props.onScheduleChange(e.target.value)}
      />
      {!props.schedule && <FieldError message="Frequency is required" />}
      <p className={css.noTopMargin}>
        Schedule is specified using unix-cron format. You can define a schedule
        so that your job runs multiple times a day, or runs on specific days and
        months.
        <LearnMoreLink href={SCHEDULE_LINK} />
      </p>
    </div>
  );
}
