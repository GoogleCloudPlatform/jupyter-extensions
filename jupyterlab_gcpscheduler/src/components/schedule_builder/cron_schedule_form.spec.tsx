/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { shallow } from 'enzyme';
import * as React from 'react';

import { simulateFieldChange, createReactWrapper } from '../../test_helpers';
import { CronScheduleBuilder } from './cron_schedule_form';

describe('CronScheduleBuilder', () => {
  const mockProps = {
    schedule: '',
    onScheduleChange: jest.fn(),
  };

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('Renders without error', async () => {
    const cronScheduleBuilder = shallow(<CronScheduleBuilder {...mockProps} />);
    expect(cronScheduleBuilder).toMatchSnapshot();
  });

  it('Triggers callback when schedule is changed', async () => {
    const cronScheduleBuilder = shallow(<CronScheduleBuilder {...mockProps} />);
    simulateFieldChange(
      createReactWrapper(cronScheduleBuilder, 'TextInput'),
      'input[name="schedule"]',
      'schedule',
      '* * * * *'
    );
    expect(mockProps.onScheduleChange).toBeCalledTimes(1);
  });

  it('Displays error when schedule is empty', async () => {
    const cronScheduleBuilder = shallow(<CronScheduleBuilder {...mockProps} />);
    simulateFieldChange(
      createReactWrapper(cronScheduleBuilder, 'TextInput'),
      'input[name="schedule"]',
      'schedule',
      ''
    );
    expect(mockProps.onScheduleChange).toBeCalledTimes(1);
    expect(cronScheduleBuilder.find('TextInput').prop('error')).toEqual(
      'Frequency is required'
    );
  });
});
