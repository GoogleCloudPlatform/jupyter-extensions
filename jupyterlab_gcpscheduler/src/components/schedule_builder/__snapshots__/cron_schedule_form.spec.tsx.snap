// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CronScheduleBuilder Renders without error 1`] = `
<div>
  <TextInput
    error="Frequency is required"
    formHelperLink="https://cloud.google.com/scheduler/docs/configuring/cron-job-schedules"
    formHelperText="Schedule is specified using unix-cron format. You can define a schedule
        so that your execution runs multiple times a day, or runs on specific
        days and months."
    hasError={true}
    label="Schedule"
    name="schedule"
    onChange={[Function]}
    value=""
  />
</div>
`;
